hrw.go

```diff
@@ -1,280 +1,124 @@
 // Package hrw implements Rendezvous hashing.
-// http://en.wikipedia.org/wiki/Rendezvous_hashing.
+// https://en.wikipedia.org/wiki/Rendezvous_hashing.
 package hrw
 
 import (
-	"encoding/binary"
-	"errors"
-	"math"
-	"reflect"
 	"sort"
 
 	"github.com/twmb/murmur3"
+	"golang.org/x/exp/constraints"
 )
 
-type (
-	// Hasher interface used by SortSliceByValue.
-	Hasher interface{ Hash() uint64 }
+// Hashable is something that can be hashed.
+type Hashable interface{ Hash() uint64 }
 
-	sorter struct {
-		l    int
-		less func(i, j int) bool
-		swap func(i, j int)
-	}
-)
-
-// Boundaries of valid normalized weights.
-const (
-	NormalizedMaxWeight = 1.0
-	NormalizedMinWeight = 0.0
-)
-
-func (s *sorter) Len() int           { return s.l }
-func (s *sorter) Less(i, j int) bool { return s.less(i, j) }
-func (s *sorter) Swap(i, j int)      { s.swap(i, j) }
-
-func distance(x uint64, y uint64) uint64 {
-	acc := x ^ y
-	// here used mmh3 64 bit finalizer
-	// https://github.com/aappleby/smhasher/blob/61a0530f28277f2e850bfc39600ce61d02b518de/src/MurmurHash3.cpp#L81
-	acc ^= acc >> 33
-	acc = acc * 0xff51afd7ed558ccd
-	acc ^= acc >> 33
-	acc = acc * 0xc4ceb9fe1a85ec53
-	acc ^= acc >> 33
-	return acc
+// Hash hashes the data in a fashion similar to the
+// [Hashable] interface. Can be used as a hash function
+// required for the HRW hashing algorithm.
+func Hash(data []byte) uint64 {
+	return murmur3.Sum64(data)
 }
 
-// Hash uses murmur3 hash to return uint64.
-func Hash(key []byte) uint64 {
-	return murmur3.Sum64(key)
+// WrapHash implements [Hashable] over already calculated hash.
+// Can be used for [Sort] and [SortWeighted] if a hash for an
+// object is already known.
+func WrapHash(h uint64) Hashable {
+	return hashableWrapper{h}
 }
 
-// Sort receive nodes and hash, and sort it by distance.
-func Sort(nodes []uint64, hash uint64) []uint64 {
-	l := len(nodes)
-	sorted := make([]uint64, l)
-	dist := make([]uint64, l)
-	for i := range nodes {
-		sorted[i] = uint64(i)
-		dist[i] = distance(nodes[i], hash)
-	}
-
-	sort.Slice(sorted, func(i, j int) bool {
-		return dist[sorted[i]] < dist[sorted[j]]
-	})
-	return sorted
+type hashableWrapper struct {
```

Comment:
Do you need a `struct` for it?

Reply:
can drop it. that is a personal preference: i like structs (even with a single field) more than custom types

Reply:
My primary concern wrt this library is speed/allocations of course. Usually not having wrapper structs is faster. If it is, then yes, we should do this. If it's not --- not worth the trouble.